lucille-core/src/main/java/com/kmwllc/lucille/indexer/OpenSearchIndexer.java

```diff
@@ -92,46 +101,47 @@ protected void sendToIndex(List<Document> documents) throws Exception {
 
-    // determine what field to use as id field and iterate over the documents
     for (Document doc : documents) {
-      Map<String, Object> indexerDoc = getIndexerDoc(doc);
+      Map<String, Object> indexerDoc = doc.asMap();
 
       // remove children documents field from indexer doc (processed from doc by addChildren method call below)
       indexerDoc.remove(Document.CHILDREN_FIELD);
 
       // if a doc id override value exists, make sure it is used instead of pre-existing doc id
       String docId = Optional.ofNullable(getDocIdOverride(doc)).orElse(doc.getId());
+      indexerDoc.put(Document.ID_FIELD, docId);
 
       // handle special operations required to add children documents
       addChildren(doc, indexerDoc);
 
-      // create new IndexRequest
-      IndexRequest indexRequest = new IndexRequest(index);
-      indexRequest.id(docId);
-      if (routingField != null) {
-        indexRequest.routing(doc.getString(routingField));
-      }
-      indexRequest.source(indexerDoc);
-
-      if (versionType != null) {
-        indexRequest.versionType(versionType);
-        if (versionType == VersionType.EXTERNAL || versionType == VersionType.EXTERNAL_GTE) {
-          // the partition doesn’t need to be included in the version. We assume the doc id is used as the
-          // kafka message key, which should guarantee that all “versions” of a given document have the
-          // same kafka message key and therefore get mapped to the same topic partition.
-          indexRequest.version(((KafkaDocument) doc).getOffset());
-        }
+      String routing = doc.getString(routingField);
+      Long versionNum = (versionType == VersionType.External || versionType == VersionType.ExternalGte)
+        ? ((KafkaDocument) doc).getOffset()
+        : null;
+
+      if (update) {
+        br.operations(op -> op
+            .update((up) -> {
+              up.index(index).id(docId);
+              if (routingField != null) up.routing(routing);
+              if (versionNum != null)  up.versionType(versionType).version(versionNum);
+              return up.document(indexerDoc);}));
+      } else {
+        br.operations(op -> op
+            .index((up) -> {
+              up.index(index).id(docId);
+              if (routingField != null) up.routing(routing);
+              if (versionNum != null)  up.versionType(versionType).version(versionNum);
+              return up.document(indexerDoc);}));
       }
-
-      // add indexRequest to bulkRequest
-      bulkRequest.add(indexRequest);
     }
-
-    BulkResponse response = client.bulk(bulkRequest, RequestOptions.DEFAULT);
-    if (response.hasFailures()) {
-      log.error(response.buildFailureMessage());
+    BulkResponse response = client.bulk(br.build());
+    if (response.errors()) {
+      for (BulkResponseItem item : response.items()) {
+        log.error("OpenSearchIndexer response has received error(s)", item.error().reason());
```

Comment:
We should discuss what is the desired behavior here @rseitz, what do you think? 

It seems like if we have an error here we should at least throw an IndexerException so that the batch is failed. Otherwise, we are recording success events for documents that weren't indexed. We aren't doing this in the current implementation of the ElasticIndexer or OpensearchIndexer but I believe that is a gap. 

Beyond that, can we take advantage of the granularity that the bulk API provides so as to SUCCEED and FAIL the correct documents within the batch? Support for that would probably have to be implemented at the base Indexer level.

Reply:
Thx for bringing this up, I was thinking about it too. An indexer subclass is supposed to fail the whole batch if any doc in that batch has an error. But OpenSearchIndexer and ElasticsearchIndexer are not doing that. I think they should both be updated to inspect the BulkResponse and throw an IndexerException if any error is encountered. I'm trying to remember if there was a reason we didn't do that during the initial implementations of these indexers, but I can't see a justification. SolrIndexer should be considered too -- solrClient.add() returns an UpdateResponse that we're currently ignoring, but as far as I can see there's nothing in there and, in the solr case, we can that any errors would be indicated as exceptions thrown by add().

It would be nice to take advantage of the granularity of the bulk API, it would be a good design ticket for the backlog, but I think we should address the immediate gap first.

Reply:
No. Logging + Throwing makes the application log more difficult to read, usually it results in the same error logged twice. It is better to just throw in this case and let the upstream caller handle logging. Just make sure that the IndexerException has the item errors passed into it such that when the exception is handled upstream it will include the information.

Reply:
Included in new commit